pathd/path_pcep_controller.c

```diff
@@ -756,7 +920,162 @@ int pcep_main_event_handler(struct thread *thread)
 
 
 /* ------------ Helper functions ------------ */
+bool is_best_pce(struct ctrl_state *ctrl_state, int pce)
+{
+	if (ctrl_state && ctrl_state->pcc[pce]) {
+		return ctrl_state->pcc[pce]->pce_opts->is_best;
+	} else {
+		return false;
+	}
+}
+int get_previous_best_pce(struct ctrl_state *ctrl_state)
+{
+	if (!ctrl_state || !ctrl_state->pcc_count)
+		return 0;
+
+	int previous_best_pce = -1;
+	struct pcc_state **pcc = &ctrl_state->pcc[0];
+	for (int i = 0; i < MAX_PCC; i++) {
+		if (pcc[i] && pcc[i]->pce_opts
+		    && pcc[i]->pce_opts->previous_best == true
+		    && pcc[i]->status != PCEP_PCC_DISCONNECTED) {
+			previous_best_pce = i;
+			break;
+		}
+	}
+	return previous_best_pce != -1 ? previous_best_pce + 1 : -1;
+}
+int calculate_best_pce(struct ctrl_state *ctrl_state)
+{
+	int best_precedence = PCE_DEFAULT_PRECEDENCE;
+	int best_pce = -1;
+	int one_connected_pce = -1;
+	int previous_best_pce = -1;
+	int step_0_best = -1;
+	int step_0_previous = -1;
+
+	if (!ctrl_state || !ctrl_state->pcc_count)
+		return 0;
+
+	struct pcc_state **pcc = &ctrl_state->pcc[0];
+
+	// Get state
+	for (int i = 0; i < MAX_PCC; i++) {
```

Comment:
Should be changed for a map...

Reply:
The problem is not the memory usage or the efficiency, it is about the complexity of having to deal with multiple types of indexes. We can go with that for now, there is more important changes, but this should be changed to a tre or a hash map soon.